core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2098,11 +2098,46 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    val n         = as.size
+    val resultArr = new Array[Any](as.size)
```

Comment:
You can use `n` here because you already forced the `Iterable` before.

Regarding the `new Array[Any]` I'm wondering if we can demand a `B: ClassTag` in the method definition, in order to avoid the cast at the end. Would the Scala compiler be able to help us here?

Reply:
Yes, the size twice was a silly mistake.

I was trying very hard not to alter the signature of the method. Making changes here has wider ranging effects than it might initially appear. A subtly wrong version that I wrote at first managed to lock up the entire build due to it being used in zio-test.

Reply:
It's not a mistake. That's why we review code. :)